scripts/polyrith_sage.py

```diff
@@ -59,15 +62,19 @@ def __init__(self, ename, evalue, message='Error in Sage communication'):
         self.message = message
         super().__init__(self.message)
 
-def parse_response(resp: str) -> str:
+def parse_response(resp: str) -> Dict[str, Any]:
     exp, data = resp.split(';', 1)
     return dict(power=int(exp), coeffs=json.loads(data))
 
 
-def evaluate_in_sage(query: str) -> str:
-    data = {'code': query}
-    headers = {'content-type': 'application/x-www-form-urlencoded'}
-    response = requests.post('https://sagecell.sagemath.org/service', data, headers=headers).json()
+def evaluate_in_sage(query: str) -> Dict[str, Any]:
+    data = urllib.parse.urlencode({'code': query}).encode('utf-8')
```

Comment:
Clarifying to my future self: `urlencode` returns an ASCII string, and that must be encoded to bytes before using it in a request. Actually, reading the documentation --- shouldn't you encode this to bytes, using e.g. `bytes(url, 'utf-8')` where (`url` is `urllib.parse.urlencode({'code': query})`)?

> For an HTTP POST request method, data should be a buffer in the standard application/x-www-form-urlencoded format. The [urllib.parse.urlencode()](https://docs.python.org/3/library/urllib.parse.html#urllib.parse.urlencode) function takes a mapping or sequence of 2-tuples and returns an ASCII string in this format. It should be encoded to bytes before being used as the data parameter.

Reply:
That's correct, I did encode the string into bytes using `.encode('utf-8')`. That is equivalent to `bytes(..., 'utf-8')` as for as I know.